soc: xtensa: esp32: adds initial support to AMP for dual core ESP32.

uLipe opened the pull request:
This PR adds the initial infrastructure to build and load a separated image to be executed only in the ESP32 APP_CPU plus its reserved resources.

It is also provides the inter processor message device driver to communicate between the two CPUs.
In this mechanism, users build the firmware using the esp32_net board, which generates a array C source file, which can be copied to the pro_cpu targeted zephyr app which uses CONFIG_ESP32_NETWORK_CORE. Building this project the esp32_net firmware is embedded in the final binary image and the pro_cpu loads it and starts the app_cpu during zephyr startup.

uLipe commented:
Anyone knows what could be the problem with the Documentation step?

mbolivar-nordic commented on samples/drivers/ipm/ipm_esp32/README.rst:
```diff
@@ -0,0 +1,61 @@
+and file moving, first of all you need to build the esp32_net firmware as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/ipm/ipm_esp32/ipm_esp32_net
+   :board: esp32_net
+   :compact:
```
You appear to be missing a `:goals:` here.

sylvioalves commented:
@cfriedt @carlescufi Would you be able to review this as well?

uLipe commented:
@cfriedt , @nashif , @carlescufi would you be able to review this one?

I would like to add the wifi and BT running on top of this.

carlocaione commented on dts/bindings/ipm/espressif,esp32-ipm.yaml:
```diff
@@ -0,0 +1,20 @@
+  interrupts:
+    required: false
+
+  ipm_channels:
+    type: int
+    required: true
+    description: number of channels
```
Is this used in the driver?

carlocaione commented on dts/xtensa/espressif/esp32.dtsi:
```diff
@@ -81,6 +86,12 @@
+		ipm0: ipm@3ffe5a30 {
+			compatible = "espressif,esp32-ipm";
+			reg = <0x3FFE5A30 0x8>;
+			ipm_channels = <4>;
```
If the driver requires channels I strongly suggest to move this driver to the MBOX API instead.

uLipe replied:
No, it has a single channel only, the initial design featured multiple channels and this property was a result.

Should I update this file?

carlocaione replied:
Yes. Remove it if you do not use the property.

cfriedt commented:
> @cfriedt , @nashif , @carlescufi would you be able to review this one?
> 
> I would like to add the wifi and BT running on top of this.

It looks OK to me. I wish I had more feedback, but I might let @carlocaione give his approval since he has already commented.

I might suggest clarifying somewhere that the "pro" CPU means "protocol", i wasn't sure about that. I assumed it was a baseband processor.

uLipe commented:
Folks just rebased and applied the @carlocaione suggestions, PTAL :)

uLipe commented:
@cfriedt any reason why the build with Clang job is failling? It says in the log the job was cancelled.

stephanosio commented:
> any reason why the build with Clang job is failling? It says in the log the job was cancelled.

It is because AWS killed the CI runner (it is a spot instance for cost saving reasons). You can just re-run the failed workflow when that happens.

uLipe commented:
@carlocaione PTAL again?

carlocaione commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+		id = 0;
+	} else {
+		id = (uint32_t )user_data;
+	}
```
Isn't this exactly the use case for MBOX instead of IPM? You are registering different callbacks on different channels.

uLipe replied:
Oh, it seems this is a leftovers from the development version, the current IPC implementation has a single callback and a single channel, the id is just a placeholder to pass IPC frame identification to the upper layer which is responsible to decode.

uLipe replied:
Oh, it seems this one is a previous commit, the actual implementation is not like that,

carlocaione commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+	
+	/* pro_cpu is responsible to initialize the lock of shared memory */
+	if(data->this_core_id == 0) {
+		irq_source = DT_IRQN(DT_NODELABEL(ipi0));
```
same here. Can you reference `ipi0` and `ipi1` from DT instead of hardcoding?

uLipe replied:
The ipi0 and ipi1 are hardwired to cpu number they belong, I don't see any advantage to remove the hardcode here.

carlocaione reviewed (changes requested):
Please squash the commits in a sensible way. Each commit is touching something introduced by a previous commit in the same PR.

uLipe commented on soc/xtensa/esp32_net/Kconfig.soc:
```diff
@@ -0,0 +1,83 @@
+	bool "Number of universally administered (by IEEE) MAC address"
+	default ESP32_UNIVERSAL_MAC_ADDRESSES_FOUR
+	help
+	  Configure the number of universally administered (by IEEE) MAC addresses.
+	  During initialization, MAC addresses for each network interface are generated or
+	  derived from a single base MAC address. If the number of universal MAC addresses is four,
```
This was imported from esp32 Kconfig file which is currently present in the upstream, I would like to handle this issue into a separate cleanup PR for both esp32, what do you think @carlocaione ?

carlocaione replied:
yup, no prob

uLipe commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+static struct esp32_ipm_data esp32_ipm_device_data;
+
+static size_t esp32_ipm_channel_data_size(uint32_t id) {
+	switch(id) {
```
Please check what is the commit you're looking at, this is not appearing into the latest commit.

Probably after I squash the commits this will be more clear

uLipe commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+{
+	switch(id) {
+
+		case CONFIG_ESP32_IPM_CHANNEL_WIFI:
```
Same here.

uLipe commented on soc/xtensa/esp32/soc.h:
```diff
@@ -31,6 +31,16 @@ static inline void esp32_clear_mask32(uint32_t v, uint32_t mem_addr)
 	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
 }
 
+static inline uint32_t esp32_core_id(void)
```
AFAIK  `arch_proc_id()` depends on correct configuration of cpu_pointer structure, which only occurs for xtensa code that enables SMP, since is not  the case here, this function will always return the core 0, so I preferred to introduce a SoC specific function that returns the core id regardless the SMP or not.

I think this is a separate issue, correct the xtensa arch code in respect of this and after that submit a new PR to deprecate or remove this function in the future.

carlocaione replied:
Yes, this needs to be fixed at arch level.

uLipe commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+	
+	data->this_core_id = esp32_core_id();
+	data->other_core_id = (data->this_core_id  == 0) ? 1 : 0;
+	data->shm = (struct esp32_ipm_memory *)DT_REG_ADDR(DT_NODELABEL(shm0));
+	data->control = (struct esp32_ipm_control *)DT_REG_ADDR(DT_NODELABEL(ipm0));
```
This is true, but since this is a software implementation of IPM with a single instance, I don't see any problem to hardcode the nodes here, this IPM won't scale beyond a single instance here, tough.

carlocaione replied:
the problem is that you are relying on the fact that the label will never change. If for some reason you are going to change `shm0` to `shm` this will break, this is why you usually do not want to hardcode labels here.

mbolivar-nordic replied:
I agree with @carlocaione ;  this is weird for multiple reasons. Instead, I think it would be cleaner to do something like this:

1. add two phandle properties to the ipm node which point to the shm and control memories
2. update the DTS file appropriately to point to the existing shm0/shm1

@carlocaione what do you think about that?

carlocaione replied:
yes, using phandle is the correct choice here IMO.

carlocaione replied:
Are you going to address these comments?

uLipe replied:
Yes, just changed locally, finding some time to test before pushing

uLipe commented on drivers/ipm/ipm_esp32.c:
```diff
@@ -0,0 +1,280 @@
+	size_t channel_size = esp32_ipm_channel_data_size(id);
+	struct esp32_ipm_data *dev_data = (struct esp32_ipm_data *)dev->data;
+
+	if(data == NULL) {
```
This part of the code does not exist anymore.

Commit:
soc: esp32_net: added esp32_net as an option

to use esp32 core 1 as a separate CPU, also
provide infrastructure to load firmware to
that CPU

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
drivers: ipm: esp32: added IPM driver

implemented by software for esp32 dual core
variants.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
samples: drivers: ipm: added esp32 sample

code to demonstrate the utilization of assymetric
dual core infrastructure based on soft-IPM implementation.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

jeremybettis commented on cmake/toolchain/espressif/target.cmake:
```diff
@@ -9,6 +9,7 @@ set(LINKER ld)
 set(BINTOOLS gnu)
 
 set(CROSS_COMPILE_TARGET_xtensa_esp32    xtensa-esp32-elf)
+set(CROSS_COMPILE_TARGET_xtensa_esp32_net  xtensa-esp32-elf)
 set(CROSS_COMPILE_TARGET_xtensa_esp32s2  xtensa-esp32s2-elf)
 set(CROSS_COMPILE_TARGET_riscv_esp32c3   riscv32-esp-elf)
```
Should these be aligned as they were before?

uLipe replied:
You are right, thanks for the remider.

carlocaione commented:
Any reason you didn't address this comment? https://github.com/zephyrproject-rtos/zephyr/pull/44645/commits/cf5cf336506bd9e00c004f985eeae7c6bcc050a3#r923912264

uLipe commented:
@carlocaione , PTAL.

Just addressed the hardcoding mixing your and @mbolivar-nordic proposals.

The shared memory now is a property from soft-IPM binding, this removes the hardcoded labels direct in the drive.

Commit:
drivers: ipm: ipm_esp32: remove hardcoded nodelabels

by making shared memory property  of IPM binding.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
soc: xtensa: esp32: fix app cpu boot

procedure by setting its PC to 0 before reset.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

uLipe commented:
@carlocaione, just pinging you again, PTAL when you have time :)

uLipe commented:
@jeremybettis PTAL again when you can?

uLipe commented:
Also, @sylvioalves , @cfriedt

sylvioalves commented on boards/xtensa/esp32_net/doc/index.rst:
```diff
@@ -0,0 +1,153 @@
+Prerequisites
+=============
+
+The ESP32 toolchain :file:`xtensa-esp32-elf` is required to build this port.
```
The content in this file needs to be updated. Check latest `boards/xtensa/esp32/doc/index.rst`.  
Currently, ESP32 toolchain is provided in Zephyr's SDK.

sylvioalves commented on boards/xtensa/esp32_net/esp32_net.yaml:
```diff
@@ -0,0 +1,13 @@
+type: mcu
+arch: xtensa
+toolchain:
+  - espressif
```
Replace to `zephyr` to allow CI testing.

cfriedt reviewed (dismissed):
LGTM

sylvioalves commented on boards/xtensa/esp32_net/esp32_net.yaml:
```diff
@@ -0,0 +1,13 @@
+  - watchdog
+  - uart
+  - pinmux
+  - nvs
```
Add the following to ignore wifi and bt testing:  
```c
testing:
  ignore_tags:
    - net
    - bluetooth
```

sylvioalves commented on cmake/toolchain/espressif/target.cmake:
```diff
@@ -8,9 +8,10 @@ set(COMPILER gcc)
-set(CROSS_COMPILE_TARGET_xtensa_esp32    xtensa-esp32-elf)
-set(CROSS_COMPILE_TARGET_xtensa_esp32s2  xtensa-esp32s2-elf)
-set(CROSS_COMPILE_TARGET_riscv_esp32c3   riscv32-esp-elf)
+set(CROSS_COMPILE_TARGET_xtensa_esp32     xtensa-esp32-elf)
```
This isn't required anymore.  
I'll submit PR removing espressif's toolchain cmake content soon.

nashif commented on soc/xtensa/esp32/soc.h:
```diff
@@ -31,6 +31,16 @@ static inline void esp32_clear_mask32(uint32_t v, uint32_t mem_addr)
 	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
 }
 
+static inline uint32_t esp_core_id(void)
```
cant you use existing `arch_proc_id()`?

uLipe replied:
Not possible yet, arch_proc_id() was developed with SMP in mind, that is it, it relies the on the SMP code to initialize the proper CPU pointers during kernel initialization. In AMP scenarios this code is not linked, and it would need to be modified.

Using it without SMP code makes the arch_proc_id() to return always CPU0

uLipe replied:
@nashif , what do you think?

sylvioalves commented on samples/drivers/ipm/ipm_esp32/ipm_esp32_net/src/main.c:
```diff
@@ -0,0 +1,37 @@
+
+	ipm_dev = DEVICE_DT_GET(DT_NODELABEL(ipm0));
+	if (!ipm_dev) {
+		printk("Failed to get IPM device.\n\r");
```
There is a missing `return` in here if it fails.

sylvioalves commented on samples/drivers/ipm/ipm_esp32/src/main.c:
```diff
@@ -0,0 +1,50 @@
+
+	ipm_dev = DEVICE_DT_GET(DT_NODELABEL(ipm0));
+	if (!ipm_dev) {
+		printk("Failed to get IPM device.\n\r");
```
Suggestion to add `return`.

sylvioalves commented on soc/xtensa/esp32_net/Kconfig.soc:
```diff
@@ -0,0 +1,75 @@
+	int
+	default 2 if ESP32_UNIVERSAL_MAC_ADDRESSES_TWO
+	default 4 if ESP32_UNIVERSAL_MAC_ADDRESSES_FOUR
+
```
There are new configs in `esp32` soc. Can you check whether it should be added in here as well?  
Example:
```c
config ESP32_PHY_MAX_WIFI_TX_POWER
	int "Max WiFi/BLE TX power (dBm)"
	range 10 20
	default 20
	help
	  Set maximum transmit power for WiFi radio. Actual transmit power for high
	  data rates may be lower than this setting.

config ESP32_PHY_MAX_TX_POWER
	int
	default ESP32_PHY_MAX_WIFI_TX_POWER
```

uLipe replied:
The support for WIFI and BLE through APP-CPU is not in place yet, we can review those configs in a next PR

sylvioalves commented on soc/xtensa/esp32_net/linker.ld:
```diff
@@ -0,0 +1,438 @@
+
+/**
+ * @file
+ * @brief Linker command/script file
```
There are lots of changes in current `esp32` linker. I suggest updating this as well.

uLipe replied:
Most of them does nott apply here, I shall remind the memory available for esp32 APP_CPU in a AMP configs is available in IRAM+DRAM only, again this can be reviewed in a new PR when adding support for flash memory

sylvioalves commented on soc/xtensa/esp32_net/soc.c:
```diff
@@ -0,0 +1,155 @@
+		: "r"(&_init_start));
+
+	/* Zero out BSS.  Clobber _bss_start to avoid memset() elision. */
+	(void)memset(&_bss_start, 0,
```
Could be replace to `z_bss_zero();`

sylvioalves commented on soc/xtensa/esp32s2/soc.h:
```diff
@@ -23,6 +23,10 @@
 
 void __esp_platform_start(void);
 
+static inline uint32_t esp_core_id(void)
```
Is this really used? It is in `esp32s2` soc.

uLipe replied:
yes, because interrupt allocator in esp32 and esp32s2 share the same code, so this function is needed here.

sylvioalves reviewed (changes requested):
There a few changes required, specially what is related to toolchain.

Commit:
boards: xtensa: esp32_net: update documentation

with latest esp32 information about the toolchain
for esp32.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
soc: xtensa: esp32_net: update kconfig

specific of soc to make it compatible to the new toolchain.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
samples: drivers: ipm_esp32: add return

when ipm driver initialization fails.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
soc: xtensa: esp32_net: fixes include paths

that were changed with zephyr prefix  after rebasing.
Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

uLipe commented:
Hmm, it seems after rebase I will need to fix some directories inclusion in the esp32_net side before request review again.

Commit:
soc: xtensa: esp32_net: update linker

and adds newlibc fix for esp32_net SoC

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
boards: esp32_net: ignore some build tests

because esp32_net has only small IRAM space

which may overflow for large samples.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

uLipe commented:
@sylvioalves , @nashif , @jeremybettis  PTAL one more time please.

uLipe commented:
Hi folks just lifting this up.

Please have a look when you have time :)

andyross reviewed (approved):
Sorry, I've been seeing this go by and figuring someone else would get to it. :)

No complaints here.  The changes are all internal to the board/soc, and the one spot where it needs to coordinate with existing frameworks is a very conventional-looking IPM driver.

Longer term, it might be good to think harder about documenting a firmer interoperability guide for AMP architectures.  It would be nice to share more tooling.  There's no reason we couldn't document spinlocks such that they could be implemented portably on the other side, for example.


(FWIW: check out lib/os/winstream.c for a simpler, higher bandwidth lockless shared memory IPC utility that might be a better fit for network applications than IPM.  It has minimal dependencies; the intent is you just copy that C file and header to the other app on the other core.)
